File: TSVPS/5S_TSVPS_RGR/5S_TSVPS_RGR.cpp
#include <iostream>
#include <chrono>
#include <fstream>
#include <vector>

using namespace std;

size_t regularSumCounter = 0;
size_t regularMulCounter = 0;
size_t fastSumCounter = 0;
size_t fastMulCounter = 0;

template<typename T = int>
void dgemmBlas(const vector<vector<T>>& matrixA,
	const vector<vector<T>>& matrixB,
	vector<vector<T>>& matrixC)
{
	if (matrixA.size() != matrixB.size())
	{
		return;
	}

	size_t matrixSize = matrixA.size();

	matrixC.clear();
	matrixC.shrink_to_fit();
	matrixC.resize(matrixSize, vector<T>(matrixSize));

	for (int i = 0; i < matrixSize; i++)
	{
		for (int j = 0; j < matrixSize; j++)
		{
			for (int k = 0; k < matrixSize; k++)
			{
				matrixC[i][j] += matrixA[i][k] * matrixB[k][j];
				regularSumCounter++;
				regularMulCounter++;
			}
		}
	}
}

#define MAT1(matrix) matrix[0][0]
#define MAT2(matrix) matrix[0][1]
#define MAT3(matrix) matrix[1][0]
#define MAT4(matrix) matrix[1][1]

template<typename T>
void copyQuarterOfMatrix(vector<vector<T*>>& matrix,
	vector<vector<T*>>& M, size_t quarterNum) {
	size_t partI = 0;
	size_t partJ = 0;

	partI = (quarterNum == 2 || quarterNum == 4) ? matrix.size() / 2 : 0;
	partJ = (quarterNum == 3 || quarterNum == 4) ? matrix.size() / 2 : 0;

	for (size_t startI = partI, i = 0; i < matrix.size() / 2; startI++, i++) {
		for (size_t startJ = partJ, j = 0; j < matrix.size() / 2; startJ++, j++)
		{
			M[i][j] = matrix[startI][startJ];
		}
	}
}

template<typename T>
void copyAllQuarterOfMatrix(vector<vector<T*>>& matrix,
	vector<vector<T*>>& M1,
	vector<vector<T*>>& M2,
	vector<vector<T*>>& M3,
	vector<vector<T*>>& M4)
{
	copyQuarterOfMatrix(matrix, M1, 1);
	copyQuarterOfMatrix(matrix, M2, 2);
	copyQuarterOfMatrix(matrix, M3, 3);
	copyQuarterOfMatrix(matrix, M4, 4);
}


template<typename T>
void fastMatrixMultiplyAction(vector<vector<T*>>& A,
	vector<vector<T*>>& B,
	vector<vector<T*>>& C) {

	size_t sumCounter = 18;
	fastMulCounter += 7 * (A.size() - 1);

	if (A.size() > 2) {
		size_t matrixSize = A.size();
		size_t matrixHalfSize = A.size() / 2;

		vector<vector<T*>> A1(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> A2(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> A3(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> A4(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));

		vector<vector<T*>> B1(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> B2(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> B3(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> B4(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));

		vector<vector<T*>> C1(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> C2(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> C3(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));
		vector<vector<T*>> C4(matrixHalfSize, vector<T*>(matrixHalfSize, nullptr));

		copyAllQuarterOfMatrix(A, A1, A2, A3, A4);
		copyAllQuarterOfMatrix(B, B1, B2, B3, B4);
		copyAllQuarterOfMatrix(C, C1, C2, C3, C4);

		fastMatrixMultiplyAction(A1, B1, C1);
		fastMatrixMultiplyAction(A2, B2, C2);
		fastMatrixMultiplyAction(A3, B3, C3);
		fastMatrixMultiplyAction(A4, B4, C4);

		fastSumCounter += sumCounter * 18 * 4;
	}
	else if (A.size() == 2) {
		auto M1 = (*MAT2(A) - *MAT4(A)) * (*MAT3(B) + *MAT4(B));
		auto M2 = (*MAT1(A) + *MAT4(A)) * (*MAT1(B) + *MAT4(B));
		auto M3 = (*MAT1(A) - *MAT3(A)) * (*MAT1(B) + *MAT2(B));
		auto M4 = (*MAT1(A) + *MAT2(A)) * *MAT4(B);
		auto M5 = *MAT1(A) * (*MAT2(B) - *MAT4(B));
		auto M6 = *MAT4(A) * (*MAT3(B) - *MAT1(B));
		auto M7 = (*MAT3(A) + *MAT4(A)) * *MAT1(B);

		*MAT1(C) = M1 + M2 - M4 + M6;
		*MAT2(C) = M4 + M5;
		*MAT3(C) = M6 + M7;
		*MAT4(C) = M2 - M3 + M5 - M7;
		fastSumCounter += sumCounter;
	}

}

template<typename T>
void fastMatrixMultiply(vector<vector<T>>& matrixA,
	vector<vector<T>>& matrixB,
	vector<vector<T>>& matrixC) {

	if (matrixA.size() != matrixB.size())
	{
		return;
	}

	size_t matrixSize = matrixA.size();

	matrixC.clear();
	matrixC.shrink_to_fit();
	matrixC.resize(matrixSize, vector<T>(matrixSize));

	vector<vector<T*>> pA(matrixSize, vector<T*>(matrixSize, nullptr));
	vector<vector<T*>> pB(matrixSize, vector<T*>(matrixSize, nullptr));
	vector<vector<T*>> pC(matrixSize, vector<T*>(matrixSize, nullptr));

	for (size_t i = 0; i < matrixSize; i++)
	{
		for (size_t j = 0; j < matrixSize; j++)
		{
			pA[i][j] = &matrixA[i][j];
			pB[i][j] = &matrixB[i][j];
			pC[i][j] = &matrixC[i][j];
		}
	}

	fastMatrixMultiplyAction(pA, pB, pC);
}

int main()
{
	ofstream out;
	out.open("out.csv", ios::out);
	if (!out.is_open())
	{
		cerr << "Unable to create output file." << endl;
		return EXIT_FAILURE;
	}
	std::cout.rdbuf(out.rdbuf());

	cout << "Size;Regular sum;Regular mul;Fast sum;Fast mul;Regular total;Fast total" << "\n";

	for (size_t matrixSize = 2; matrixSize <= 512; matrixSize *= 2) {
		regularSumCounter = 0;
		regularMulCounter = 0;
		fastSumCounter = 0;
		fastMulCounter = 0;

		vector<vector<long long>> A(matrixSize, vector<long long>(matrixSize));
		vector<vector<long long>> B(matrixSize, vector<long long>(matrixSize));
		vector<vector<long long>> C(matrixSize, vector<long long>(matrixSize));

		for (size_t i = 0; i < matrixSize; i++)
		{
			for (size_t j = 0; j < matrixSize; j++)
			{
				A[i][j] = static_cast<size_t>(pow(-1, i + j));
				B[i][j] = i + j;
			}
		}

		dgemmBlas(A, B, C);
		fastMatrixMultiply(A, B, C);

		cout
			<< matrixSize << ";"
			<< regularSumCounter << ";"
			<< regularMulCounter << ";"
			<< fastSumCounter << ";"
			<< fastMulCounter << ";"
			<< regularSumCounter + regularMulCounter << ";"
			<< fastSumCounter + fastMulCounter << endl;
	}

	return EXIT_SUCCESS;
}

File: TSVPS/5S_TSVPS_RGR/5S_TSVPS_RGR.py
def getMQuarter(A, num):
    if isinstance(A, list):
        matrixSize = len(A)
        quarterSize = int(matrixSize / 2)
        Q = [[0 for _ in range(quarterSize)] for _ in range(quarterSize)]

        if num == 1:
            startI = 0
            endI = quarterSize
            startJ = 0
            endJ = quarterSize
        elif num == 2:
            startI = 0
            endI = quarterSize
            startJ = quarterSize
            endJ = matrixSize
        elif num == 3:
            startI = quarterSize
            endI = matrixSize
            startJ = 0
            endJ = quarterSize
        elif num == 4:
            startI = quarterSize
            endI = matrixSize
            startJ = quarterSize
            endJ = matrixSize
        else:
            raise RuntimeError("Quarter is greater than 4 or less then 0")

        for iQ, i in zip(range(quarterSize), range(startI, endI)):
            for jQ, j in zip(range(quarterSize), range(startJ, endJ)):
                Q[iQ][jQ] = A[i][j]
        if len(Q) == 1 and len(Q[0]) == 1:
            return Q[0][0]

        return Q

    return A


def minus(A, B, counters):
    if isinstance(A, list) and isinstance(B, list):
        Q = [[0 for _ in range(len(A))] for _ in range(len(A))]

        for i in range(len(Q)):
            for j in range(len(Q)):
                Q[i][j] = A[i][j] - B[i][j]
                counters['sum'] += 1
        return Q

    counters['sum'] += 1
    return A - B


def plus(A, B, counters):
    if isinstance(A, list) and isinstance(B, list):
        Q = [[0 for _ in range(len(A))] for _ in range(len(A))]

        for i in range(len(Q)):
            for j in range(len(Q)):
                Q[i][j] = A[i][j] + B[i][j]
                counters['sum'] += 1
        return Q

    counters['sum'] += 1
    return A + B


def mul(A, B, counters):
    if isinstance(A, list) and isinstance(B, list):
        return fastMatrixMultiply(A, B, counters)

    counters['mul'] += 1
    return A * B


def fastMatrixMultiply(A, B, counters):
    M1 = mul(
        minus(getMQuarter(A, 2), getMQuarter(A, 4), counters),
        plus(getMQuarter(B, 3), getMQuarter(B, 4), counters),
        counters
    )
    M2 = mul(
        plus(getMQuarter(A, 1), getMQuarter(A, 4), counters),
        plus(getMQuarter(B, 1), getMQuarter(B, 4), counters),
        counters
    )
    M3 = mul(
        minus(getMQuarter(A, 1), getMQuarter(A, 3), counters),
        plus(getMQuarter(B, 1), getMQuarter(B, 2), counters),
        counters
    )
    M4 = mul(
        plus(getMQuarter(A, 1), getMQuarter(A, 2), counters),
        getMQuarter(B, 4),
        counters
    )
    M5 = mul(
        getMQuarter(A, 1),
        minus(getMQuarter(B, 2), getMQuarter(B, 4), counters),
        counters
    )
    M6 = mul(
        getMQuarter(A, 4),
        minus(getMQuarter(B, 3), getMQuarter(B, 1), counters),
        counters
    )
    M7 = mul(
        plus(getMQuarter(A, 3), getMQuarter(A, 4), counters),
        getMQuarter(B, 1),
        counters
    )

    C1 = plus(minus(plus(M1, M2, counters), M4, counters), M6, counters)
    C2 = plus(M4, M5, counters)
    C3 = plus(M6, M7, counters)
    C4 = minus(plus(minus(M2, M3, counters), M5, counters), M7, counters)

    if isinstance(C1, list):
        C = [[] for _ in range(len(A))]

        indexC = 0
        for i in range(int(len(C) / 2)):
            C[indexC] = [*C1[i], *C2[i]]
            indexC += 1

        for i in range(int(len(C) / 2)):
            C[indexC] = [*C3[i], *C4[i]]
            indexC += 1
    else:
        C = [[C1, C2], [C3, C4]]

    counters['sum'] += 18
    counters['mul'] += 7

    return C


def dgemmBlas(A, B, counters):
    C = [[0 for _ in range(len(A))] for _ in range(len(A))]

    for i in range(len(A)):
        for j in range(len(A)):
            for k in range(len(A)):
                C[i][j] += A[i][k] * B[k][j]
                counters['sum'] += 1
                counters['mul'] += 1
    return C


def main():
    matrixSize = 2
    with open('out.csv', 'w+') as file:
        file.write("Size;Regular +;Regular *;Fast +;Fast *;Regular total;Fast total\n")
        while matrixSize <= 512:
            print("Processing: ", matrixSize, "*", matrixSize, "...", sep=" ")
            A = [[pow(-1, x + y) for x in range(matrixSize)] for y in range(matrixSize)]
            B = [[x + y for x in range(matrixSize)] for y in range(matrixSize)]

            regCounters = dict()
            regCounters['sum'] = 0
            regCounters['mul'] = 0

            fastCounters = dict()
            fastCounters['sum'] = 0
            fastCounters['mul'] = 0

            dgemmBlas(A, B, regCounters)
            fastMatrixMultiply(A, B, fastCounters)

            outStr = str(matrixSize) + " x " + str(matrixSize) + ";" \
                + str(regCounters['sum']) + ";" \
                + str(regCounters['mul']) + ";" \
                + str(fastCounters['sum']) + ";" \
                + str(fastCounters['mul']) + ";" \
                + str(regCounters['sum'] + regCounters['mul']) + ";" \
                + str(fastCounters['sum'] + fastCounters['mul']) + ";" \
                + "\n"

            file.write(outStr)
            file.flush()

            matrixSize *= 2


if __name__ == '__main__':
    main()
